test/include/constraints/RegularPenaltyEqualValueConstraint.h

```diff
@@ -0,0 +1,31 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MortarConstraint.h"
+
+/**
+ * Constrain the value of a variable to be the same on both sides of an
+ * interface using a generalized force stemming from a penalty-based enforcement.
+ */
+class RegularPenaltyEqualValueConstraint : public MortarConstraint
```

Comment:
Now that we'll have the non-AD mortar object, it might make sense to use some templates, which may save us some a bit of code duplication and leverage the more accepted `PenaltyEqualValueConstraint` and `ADEqualValueConstraint` pair.

Reply:
This is a test object though. I am happy to move this to the framework if you like.

Reply:
Thinking more about it, I think the non-AD version should in most cases be a superior alternative. So if I were you I would always use this non-AD version. Anyways, it's your call. Let me know where you wanna put this, and if you want to have this in the framework, just let me know how you want to name this object and I can template this on is_ad.

Reply:
> I think the non-AD version should in most cases be a superior alternative

That can definitely be true. But we could always couple this object with other physics, so having an AD object available may not be that bad. I'd vote for having a templated object in the framework. 

@lindsayad, what do you think?

Reply:
In this case I don't think there would be much reduction in duplication as the only code that is shared is the penalty value parameter and the residual code.

As to superiority, that is hard to argue at this point. In the future if we AD-ify the mortar mesh geometry and there are large mesh displacements, then that may no longer be true

Reply:
Ohhh, AD mortar mesh, that sounds super fun. And ya, AD will be much more useful in that case.

So it sounds like there are arguments either way. I guess I'll leave this as a test object as is then. The cached jacobian for P-S and S-P is the major change in this PR anyways.

Reply:
I would say that if you think this object is superior (and I would agree especially at this time) then it should go into the framework instead of go in as a test object

Reply:
I'm fine any choice. But I've see residual computations (and parameters) diverge in modules and apps.

Reply:
that's true but conceptually this object should always be pretty simple. There shouldn't be any bells and whistles being added for penalty-based continuity enforcement

Reply:
Stuff may inherit from it to provide particular directional behavior, damping, etc.